www/src/components/use-localized-strings.js

```diff
@@ -0,0 +1,51 @@
+import React from "react"
+import { defaults } from "lodash-es"
+import { useStaticQuery, graphql } from "gatsby"
+import { LocaleContext } from "./layout"
+import { defaultLang } from "../util/i18n"
+
+export default function useLocalizedStrings(component) {
+  const currentLocale = React.useContext(LocaleContext)
+
+  const { allStringsYaml } = useStaticQuery(graphql`
```

Comment:
This is the only way I've found to make this work so far, but I don't like it.

Usually, you just provide whatever i18n library you have with a set of key-value pairs, and it takes care of it. But since we're pulling in translations from different repos, they have to be sourced, which means the only way (that I know of) to reference the content is through a GraphQL query, which means listing all the individual keys and values.

This means whenever you add a new text string, you have to reference the key three times: once in `strings/components.yml`, once in this file, and once in the component itself. It's tedious and error prone.

It feels like you should be able to limit the query to whatever file you're working on, e.g.:

```
query($locale: String) {
  stringsYaml(filter: { locale: $locale }) {
    nodes {
      docsTableOfContents { title }
    }
  }
}
```

But we can't have variables in useStaticQuery, so no way to limit to the current locale.

Reply:
Are you able to use a flat yaml structure and source key/value pairs, rather than this structure? e.g. a structure like this:

```yaml
Strings:
    -
        key: SIDEBAR_LABEL_EXPAND
        value: Expand
    -
        key: SIDEBAR_LABEL_COLLAPSE
        value: Collapse
```

And then you can just query :

```graphql
query {
      allStringsYaml {
        nodes {
          Strings {
            key
            value
          }
}
```

...or something?

Reply:
Agreed that a flat file seems to be the most common way that i18n libraries do things and should be the way to go (though some do have support for [namespaces!](https://www.i18next.com/principles/namespaces)).

Turns out that the issue of "Maps in GraphQL" is a [pretty controversial one](https://github.com/graphql/graphql-spec/issues/101) (the example is also about i18n!) @TylerBarnes brought up the idea of getting the files using `getNode` in `gatsby-node` and passing that into page context. Which option would be better?

Reply:
The nice thing about using page context is you wouldn't be loading extra data, though I'm not sure how much it matters in the context of this project (not sure how many translations there are). You could do a hybrid between that suggestion and @ascorbic's suggestion where you pass the relevant strings to page context and pass them into the hook:

```jsx
const { localizationStrings } = data.pageContext
const { t } = useLocalizedStrings(currentLocale, localizationStrings);
return <button>{t('SIDEBAR_LABEL_COLLAPSE')}</button>
```

Reply:
I don't think there should be that many translations to where it would cause performance issues. What do you mean "loading extra data"?

Reply:
@tesseralis I meant loading strings that weren't relevant to the current page but I just realized it might be difficult to know that beforehand anyway. @ascorbic's way is probably the path of least resistance, especially if the amount of translations wouldn't cause performance issues.

Reply:
@freiksenet, @sidharthachatterjee and I went off on a long thread about this under your message on Slack. One option I thought of would be to split it into static strings used in components (such as button labels etc) and keep page content like the actual tutorials separate. That would make it easier to do the key/value thing for the static strings as there's far less data. @TylerBarnes's idea of passing them in page context might be even better. You'd then load the page content in the page query as usual. As it's one page, you could potentially even load every locale in the query if you want to be able to dynamically switch. Is the plan to build separate pages for each locale, or a single page and dynamically load the translations? I think the former is marginally better, as you then get SSR for every language. Is there a best practice for this?

Reply:
The actual pages are translated separate from the UI text strings, and in fact, are already [accounted for](https://build-57f60042-fb09-471c-9527-4479f5ba2c36.gtsb.io/id/tutorial/)! Yes, we're creating a new page for each language, so we get all that SSR built-in.

I'm thinking of using the pageContext approach too, and should have an updated PR using it today.

Reply:
I’m running into an issue where the page won't update after changing my content. E.g. I'll change "Table of Contents" to "Table of Elements" but the change won't be reflected in the doc unless I clear the cache. I think it might be related to #11691? If so that would put a damper on our plans to use `pageContext`.

Reply:
Great that you're way ahead of me on the handling of actual pages. Others will know better than me, but it certainly does sound like #16477 could be the issue. What you could do is rather than passing the whole lot in context, you just pass the locale name and then do the query in the pageQuery, filtering by locale. That would mean that the context wouldn't change, and changes to translations should be picked up in the page query.

One concern I do have though is that this is quite wasteful, as it's keeping all of the strings in the page data for every page, so you need to load it again whenever you navigate. I wonder if it's better to go the other way, and rather than passing the specific locale to the page, you load all of the locales and put them in context in a `wrapRootElement` function. Your custom hook could then use that root context, which has all of the locales, and which would be preserved during page navigations (I think: is that right?). Even though the initial load is larger because of it holding all of the locales, this should be more than made up for by the fact it doesn't need to load them in every page data.

Reply:
I was wondering that as well. Does `wrapRootElement` have access to page/static queries?

Another thing I was thinking of was putting it in `wrapPageElement` through `gatsby-plugin-layout`. Since that function has access to `location`, it can detect the locale from the path and get the right strings from that.

Reply:
I don't think it has access to page queries, but static queries should be fine if it wraps it in a component that has the static query